tiledb/api/src/query_arrow/arrow.rs

```diff
@@ -0,0 +1,520 @@
+    ArrowTypeNotSupported(adt::DataType),
+    #[error("DataFusion does not support multi-value cells.")]
+    InvalidMultiCellValNum,
+    #[error("The TileDB Datatype '{0}' is not supported by DataFusion")]
+    UnsupportedTileDBDatatype(Datatype),
+    #[error("Variable-length datatypes as list type elements are not supported by TileDB")]
+    UnsupportedListVariableLengthElement,
+}
+
+pub type Result<T> = std::result::Result<T, Error>;
+
+/// ConversionMode dictates whether certain conversions are allowed
+pub enum ConversionMode {
+    /// Only allow conversions that are semantically equivalent
+    Strict,
+    /// Allow conversions as long as the physical type is maintained.
+    Relaxed,
+}
+
+pub struct ToArrowConverter {
+    mode: ConversionMode,
+}
+
+impl ToArrowConverter {
+    pub fn strict() -> Self {
+        Self {
+            mode: ConversionMode::Strict,
+        }
+    }
+
+    pub fn physical() -> Self {
+        Self {
+            mode: ConversionMode::Relaxed,
+        }
+    }
+
+    pub fn convert_datatype(
+        &self,
+        dtype: &Datatype,
+        cvn: &CellValNum,
+        nullable: bool,
+    ) -> Result<adt::DataType> {
+        if let Some(arrow_type) = self.default_arrow_type(dtype) {
+            self.convert_datatype_to(dtype, cvn, nullable, arrow_type)
+        } else {
+            Err(Error::NoDefaultArrowType(*dtype))
+        }
+    }
+
+    pub fn convert_datatype_to(
+        &self,
+        dtype: &Datatype,
+        cvn: &CellValNum,
+        nullable: bool,
+        arrow_type: adt::DataType,
+    ) -> Result<adt::DataType> {
+        if matches!(arrow_type, adt::DataType::Null) {
+            return Err(Error::InvalidTargetType(arrow_type));
+        }
+
+        if arrow_type.is_primitive() {
+            let width = arrow_type.primitive_width().unwrap();
+            if width != dtype.size() as usize {
+                return Err(Error::PhysicalSizeMismatch(*dtype, arrow_type));
+            }
+
+            if cvn.is_single_valued() {
+                return Ok(arrow_type);
+            } else if cvn.is_var_sized() {
+                let field =
+                    Arc::new(adt::Field::new("item", arrow_type, nullable));
+                return Ok(adt::DataType::LargeList(field));
+            } else {
+                // SAFETY: Due to the logic above we can guarantee that this
+                // is a fixed length cvn.
+                let cvn = cvn.fixed().unwrap().get();
+                if cvn > i32::MAX as u32 {
+                    return Err(Error::CellValNumOutOfRange(cvn));
+                }
+                let field =
+                    Arc::new(adt::Field::new("item", arrow_type, nullable));
+                return Ok(adt::DataType::FixedSizeList(field, cvn as i32));
+            }
+        } else if matches!(arrow_type, adt::DataType::Boolean) {
+            if !cvn.is_single_valued() {
+                return Err(Error::RequiresSingleValued(arrow_type));
```

Comment:
How come this can't also be a list?
Is that a limitation of arrow?

Reply:
To be honest, I have no idea. I'm guessing I wrote it that way because I was focused on something else and didn't want context shift. Totes agree that I should go figure out if Arrow does support it and add that if so.